homeassistant/components/tradfri/base_class.py

```diff
@@ -44,103 +37,53 @@ async def wrapper(command: Command | list[Command]) -> None:
 
 
-class TradfriBaseClass(Entity):
-    """Base class for IKEA TRADFRI.
+class TradfriBaseEntity(CoordinatorEntity):
+    """Base Tradfri device."""
 
-    All devices and groups should ultimately inherit from this class.
-    """
-
-    _attr_should_poll = False
+    coordinator: TradfriDeviceDataUpdateCoordinator
 
     def __init__(
         self,
-        device: Device,
-        api: Callable[[Command | list[Command]], Any],
+        device_coordinator: TradfriDeviceDataUpdateCoordinator,
         gateway_id: str,
+        api: Callable[[Command | list[Command]], Any],
     ) -> None:
         """Initialize a device."""
-        self._api = handle_error(api)
-        self._attr_name = device.name
-        self._device: Device = device
-        self._device_control: BlindControl | LightControl | SocketControl | SignalRepeaterControl | AirPurifierControl | None = (
-            None
-        )
-        self._device_data: Socket | Light | Blind | AirPurifier | None = None
-        self._gateway_id = gateway_id
-
-    async def _async_run_observe(self, cmd: Command) -> None:
-        """Run observe in a coroutine."""
-        try:
-            await self._api(cmd)
-        except PytradfriError as err:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed, trying again", exc_info=err)
-            self._async_start_observe()
+        super().__init__(device_coordinator)
 
-    @callback
-    def _async_start_observe(self, exc: Exception | None = None) -> None:
-        """Start observation of device."""
-        if exc:
-            self._attr_available = False
-            self.async_write_ha_state()
-            _LOGGER.warning("Observation failed for %s", self._attr_name, exc_info=exc)
-        cmd = self._device.observe(
-            callback=self._observe_update,
-            err_callback=self._async_start_observe,
-            duration=0,
-        )
-        self.hass.async_create_task(self._async_run_observe(cmd))
-
-    async def async_added_to_hass(self) -> None:
-        """Start thread when added to hass."""
-        self._async_start_observe()
+        self._gateway_id = gateway_id
 
-    @callback
-    def _observe_update(self, device: Device) -> None:
-        """Receive new state data for this device."""
-        self._refresh(device)
+        self._device: Device = device_coordinator.data
 
-    def _refresh(self, device: Device, write_ha: bool = True) -> None:
-        """Refresh the device data."""
-        self._device = device
-        self._attr_name = device.name
-        if write_ha:
-            self.async_write_ha_state()
+        self._device_id = self._device.id
+        self._api = handle_error(api)
+        self._attr_name = self._device.name
 
+        self._attr_unique_id = f"{self._gateway_id}-{self._device.id}"
 
-class TradfriBaseDevice(TradfriBaseClass):
-    """Base class for a TRADFRI device.
+    def _refresh(self) -> None:
```

Comment:
If all children must implement this is we can make it an abstractmethod.
```suggestion
    @abstractmethod
    @callback
    def _refresh(self) -> None:
```